### ISSUE TITLE
Passing a raw Model instance to `build_callables` leads to silent incorrect results

### ISSUE THREAD

**Issue by richardotis** [pycalphad/pycalphad#208]
Compare the following on `develop`:

```python
from pycalphad import Database, Model, calculate, variables as v

# Calculation Settings
dbf = Database('alfe_sei.TDB')
comps = ['AL', 'FE', 'VA']
phase_name = 'B2_BCC'
output = 'CPM'
endmember = [[0.5, 0.5, 0.5, 0.5, 1.0]]
temp_range = [300, 1000]

mod = Model(dbf, comps, phase_name)
contributions = mod.models.copy()
results = {}

for contribution, value in contributions.items():
    mod.models.clear()
    mod.models[contribution] = value
    result = calculate(dbf, comps, phase_name,
                       model=mod, output=output, points=endmember, T=(temp_range[0], temp_range[1], 5), P=1e5)
    if (getattr(result, output) == 0).all():
        continue
    results[contribution] = getattr(result, output)
```
The `build_callables` step will incorrectly generate fresh `Model` instances, leading to incorrect results.

For the intended behavior, one must explicitly pass a dict:
```python
from pycalphad import Database, Model, calculate, variables as v

# Calculation Settings
dbf = Database('alfe_sei.TDB')
comps = ['AL', 'FE', 'VA']
phase_name = 'B2_BCC'
output = 'CPM'
endmember = [[0.5, 0.5, 0.5, 0.5, 1.0]]
temp_range = [300, 1000]

mod = Model(dbf, comps, phase_name)
contributions = mod.models.copy()
results = {}

for contribution, value in contributions.items():
    mod.models.clear()
    mod.models[contribution] = value
    result = calculate(dbf, comps, phase_name,
                       model={phase_name: mod}, output=output, points=endmember, T=(temp_range[0], temp_range[1], 5), P=1e5)
    if (getattr(result, output) == 0).all():
        continue
    results[contribution] = getattr(result, output)
```

**Comment by bocklund** [comment]
Here's a proposed solution for `build_callables`, which could help address the complex logic and edge cases that are in this, #212 and #207 

`build_callables` currently returns a dict of with keys: `['massfuncs', 'massgradfuncs', 'masshessfuncs', 'callables', 'grad_callables', 'hess_callables', 'internal_cons', 'internal_jac', 'internal_cons_hess', 'mp_cons', 'mp_jac', 'state_variables', 'model', 'phase_records']`, which all map `{phase_name: value}` (except state_variables). So many things.

Here are my proposed changes:
1. build_callables does not build `Model` instances. A `models` argument containing a `{phase_name: Model instance}` dict becomes a required argument. The `phases` argument goes away because we build callables for the phases in the models dict.
2. conditions are only used to determine whether or not Hessians are built and to build the state variables.
    I. The caller of `build_callables` should know whether Hessians are required. `build_callables` should not have to know what conditions require Hessians to be built and which do not.
    II. Logic to build state variables from Model instances should be broken into another function
    III. Callables should be built with the required state variables from (II), logic for whether the conditions specify the state variables should occur outside `build_callables` (before or after doesn't matter, because the Models are already instantiated). 
    IV. By (I) and (II), conditions are no longer required input.
3. Instead of returning a dict of dicts, we return a tuple of `(dict({phase_name: PhaseRecord()}), state_variables)` 
4. callables should not be passed. For higher level functions, if callables are passed, then we should not need to call build_callables at all (state variables, etc. can all be determined outside)

The new signature becomes

```python
build_callables(dbf, comps, models, parameters=None, output='GM', 
                build_gradients=True, build_hessians=False, verbose=False)
```

**Comment by richardotis** [comment]
I am in tentative agreement with the proposed changes, with the caveat that I'd like some sense of how this would complicate the logic in the calling functions. Part of the reason `build_callables` became so complicated was to reduce repetitive logic in various places around the core.

**Comment by bocklund** [comment]
The current code I've sketched out splits up building callables into `build_callables` (which does only that) and `build_phase_records` which 
1. builds the constraints using the conditions
2. builds callables if they are not passed (explicitly does not use conditions)
3. combines callables and constraints to build the phase records

I think it makes a little more sense to do it this way because the callables, once built, can be reused no matter the conditions, but the conditions might need to be rebuilt between equilibrium calls (even if the same callables are used) because the conditions might change. If my understanding of the current code is correct, even a change of `v.X('NI')` as a condition instead of `v.X('AL')` in the Al-Ni binary would mean that the constraints (and therefore the PhaseRecords need to be rebuilt).

One thing I'm curious about is building the constraints based on the conditions. It seems like we only want to do that if we are doing a minimization and the output is Gibbs energy. I don't think we need to hard code that in, but I'm wondering if the code path for building PhaseRecords (and the PhaseRecords themselves) need to include the constraints. What do you think?

**Comment by richardotis** [comment]
Minimization is required any time there's a call to `equilibrium`. You could avoid it by checking for independent state variables as conditions only, but even then you need to know if a minimization is planned by the calling function.

My opinion is it's a micro-optimization for `calculate`-only calls which would require a lot of additional logic to implement correctly. It's better to just always build the constraints.

**Comment by bocklund** [comment]
Closed by #214
